python/python/tests/test_pubsub.py

```diff
@@ -0,0 +1,2016 @@
+        else:
+            result[ClusterClientConfiguration.PubSubChannelModes.Sharded] = channel
+    return result
+
+
+def create_pubsub_subscription(
+    cluster_mode,
+    channels_and_patterns: Dict[PubSubChannelModes, Set[str]],
+    callback=None,
+    context=None,
+):
+    channels = create_channels_and_patterns(cluster_mode, channels_and_patterns)
+    if cluster_mode:
+        return ClusterClientConfiguration.PubSubSubscriptions(
+            channels_and_patterns=channels,
+            callback=callback,
+            context=context,
+        )
+    return RedisClientConfiguration.PubSubSubscriptions(
+        channels_and_patterns=channels, callback=callback, context=context
+    )
+
+
+def new_message(
+    msg: Union[RedisClient.PubSubMsg, RedisClusterClient.PubSubMsg], context: Any
+):
+    received_messages: List[RedisClient.PubSubMsg] = context
+    received_messages.append(msg)
+
+
+@pytest.mark.asyncio
+class TestPubSub:
+    @pytest.mark.parametrize("cluster_mode", [True, False])
+    @pytest.mark.parametrize(
+        "method", [MethodTesting.Async, MethodTesting.Sync, MethodTesting.Callback]
+    )
+    async def test_pubsub_exact_happy_path(
+        self,
+        request,
+        cluster_mode: bool,
+        method: MethodTesting,
+    ):
+        """
+        Tests the basic happy path for exact Pub/Sub functionality.
+
+        This test covers the basic pub/sub flow using three different methods:
+        Async, Sync, and Callback. It verifies that a message published to a
+        specific channel is correctly received by a subscriber.
+        """
+        channel = get_random_string(10)
+        message = get_random_string(5)
+        publish_response = 1 if cluster_mode else OK
+
+        callback, context = None, None
+        callback_messages: List[
+            Union[RedisClient.PubSubMsg, RedisClusterClient.PubSubMsg]
+        ] = []
+        if method == MethodTesting.Callback:
+            callback = new_message
+            context = callback_messages
+
+        pub_sub = create_pubsub_subscription(
+            cluster_mode,
+            {PubSubChannelModes.Exact: {channel}},
+            callback=callback,
+            context=context,
+        )
+
+        publishing_client, listening_client = await create_clients_with_subscription(
+            request, cluster_mode, pub_sub
+        )
+
+        assert await publishing_client.publish(message, channel) == publish_response
+        # allow the message to propagate
+        await asyncio.sleep(1)
+
+        pubsub_msg = await get_message_by_method(
+            method, listening_client, callback_messages, 0
+        )
+
+        assert pubsub_msg.message == message
+        assert pubsub_msg.channel == channel
+        assert pubsub_msg.pattern is None
+
+        await check_no_messages_left(method, listening_client, callback_messages, 1)
+        if cluster_mode:
```

Comment:
Do this in a `finally` block. Without it, a failure in one test cause subsequent tests to fail too.

Reply:
nice, thanks!